library/core/src/str/lossy.rs

```diff
@@ -51,19 +51,43 @@ impl<'a> Iterator for Utf8LossyChunksIter<'a> {
         }
 
         const TAG_CONT_U8: u8 = 128;
-        fn safe_get(xs: &[u8], i: usize) -> u8 {
-            *xs.get(i).unwrap_or(&0)
+
+        /// Gets the byte at `current`, returning zero if `current` is past `end`
+        ///
+        /// # Safety
+        ///
+        /// `current` must be a valid pointer to an initialized byte
+        ///
+        unsafe fn try_get(current: *const u8, end: *const u8) -> u8 {
+            // SAFETY: current is a valid pointer
+            if current < end { unsafe { *current } } else { 0 }
+        }
+
+        /// Checks if the byte at `current` bitand 192 is `TAG_CONT_U8`, returning
+        /// false if not or if `current` is past `end`
+        ///
+        /// # Safety
+        ///
+        /// `current` must be a valid pointer to an initalized byte
+        ///
+        unsafe fn shouldnt_continue(current: *const u8, end: *const u8) -> bool {
+            // SAFETY: current is a valid pointer
+            unsafe { current < end && *current & 192 != TAG_CONT_U8 }
         }
 
-        let mut i = 0;
-        let mut valid_up_to = 0;
-        while i < self.source.len() {
-            // SAFETY: `i < self.source.len()` per previous line.
+        let length = self.source.len();
+        let mut current = self.source.as_ptr();
+        // SAFETY: current + length is one past the end of the allocation
+        let (start, end, mut valid_up_to) = unsafe { (current, current.add(length), current) };
+
+        while current < end {
+            // SAFETY: `current < end` per previous line.
             // For some reason the following are both significantly slower:
             // while let Some(&byte) = self.source.get(i) {
             // while let Some(byte) = self.source.get(i).copied() {
-            let byte = unsafe { *self.source.get_unchecked(i) };
```

Comment:
This PR as-written adds a a lot of `unsafe`. I do not think that we should do that without sufficient justification.

How much of a perf impact do we see on Miri backtrace printing and the standard library benchmarks from only adjusting these lines? The specific pattern we want to avoid is this: https://github.com/rust-lang/miri/blob/8fdb720329d7674a878a8252fe4b79ef93d6ffec/bench-cargo-miri/slice-get-unchecked/src/main.rs#L8-L9
```rust
    while i < x.len() {
        let _element = unsafe { *x.get_unchecked(i) };
```

I'm not completely opposed the pervasive sort of changes that you've implemented in the rest of this function, but we need benchmarking that makes the case for those changes, as opposed to just tweaking the ASCII fast path.